### PR TITLE
Meff Attn Bias 

### PR DESCRIPTION
**Pull request by drisspg**
# Summary

### Review Points
- Automatically pad tensors to create aligned masks when seqlen_kv is not multiple of 16. This will cause memory spike ~ 2 * attn_mask size which could in theory be big.  At appears though that doing this + mem_eff is faster than no_pad + math. SO seems to be worth it
- Using expand to view the attn_mask in 4d. This is a little different to how we enforce q,k,v to be viewed in 4d prior to calling. Also not supprint b*n_heads, seq_lenq, seq_lenkv case.
- Should enable, #96099

### Profiling
I ran a bunch of comparisons between sdpa.MATH and sdp.MemEffAttention.  I added a attn_bias of shape (1, 1, seqlen_q, seqln_k). For these experiments seqlen_q == seqlen_k. These were all ran on an a100 80gb gpu.
Configs:
```
    # Run a bunch of experiments
    batch_sizes = [8, 16, 32]
    num_heads = [16, 32]
    max_seq_lens = [15, 64, 128, 512, 555, 1024]
    embed_dims = [32, 64, 128]
    dtypes = [torch.float16, torch.bfloat16, torch.float32]
    pad_percentages = [None]
    backends = [SDPBackend.EFFICIENT_ATTENTION, SDPBackend.MATH]
    run_backward = True
    attn_mask = True
```
    
   The function calls `sdpa(input**).sum().backward()`.
   
   I calculated the geomean speedup of the efficient attention path of the math path for all these configs:
   `Geomean Speedup: 1.977`
   
An example comparision with batchsize = 8, num_heads = 32, embed_dim = 64, and dtype = torch.float16:
attn_mask_compare_bsz_8_num_heads_32_embed_dim_64_dtype_fp16

 This was done using the current state of the branch where we force alignment of mask when the last dim is not divisible by 16, which shows up in seq_len = 15 and 555 case.

The full data can be found here:

   
[attn_mask_sweep.csv](https://github.com/pytorch/pytorch/files/11962399/attn_mask_sweep.csv)

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: aten/src/ATen/native/transformers/attention.cpp [hunk 1]
```diff
@@ -614,16 +660,19 @@ Tensor scaled_dot_product_attention(
       bool compute_logsumexp =
           (query_.requires_grad() || key.requires_grad() ||
            value.requires_grad());
+      if(attn_mask.has_value()){
+        attn_mask = pad_bias(attn_mask.value());
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/attention.cpp:664]: I think we should to move the expand as 4d view out here

File: aten/src/ATen/native/transformers/cuda/mem_eff_attention/pytorch_utils.h [hunk 2]
```diff
@@ -30,32 +30,33 @@ inline at::Tensor get_bias_4d_view(
...
-      TORCH_CHECK(bias.size(0) == batch_sz * n_heads);
-      return bias.view({batch_sz, n_heads, n_queries, n_keys});
-    case 4: // (batch_sz, n_heads, n_queries, n_keys) - do nothing
-      TORCH_CHECK(bias.size(0) == batch_sz);
-      TORCH_CHECK(bias.size(1) == n_heads)
-      return bias;
-    default:
-      TORCH_CHECK(false, "bias can only have ndims in {2, 3, 4}");
-  }
+    return bias.expand({batch_sz, n_heads, n_queries, n_keys});
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/mem_eff_attention/pytorch_utils.h:33]: see other comment

File: torch/_meta_registrations.py [hunk 3]
```diff
@@ -3969,8 +3971,20 @@ def meta__scaled_dot_product_efficient_backward(
...
+        assert isinstance(attn_bias, TensorLike)
+        nominal_sizes = attn_bias.size()
+        last_dim_size = nominal_sizes[-1]
+        alignTo = 16
+        nominal_sizes[-1] = alignTo * ((last_dim_size + alignTo - 1) / alignTo)
+        grad_bias = torch.empty(
+            nominal_sizes,
+            dtype=attn_bias.dtype,
+            device=attn_bias.device,
+        )[:, :, :, :last_dim_size]
```
Comment (drisspg) [inline torch/_meta_registrations.py:3985]: this can be wrong unless we move the 4d expand to the top level sdpa where it is composite

**Comment by drisspg** [comment]
I ran a bunch of comparisons between sdpa.MATH and sdp.MemEffAttention.  I added a attn_bias of shape (1, 1, seqlen_q, seqln_k). For these experiments seqlen_q == seqlen_k. These were all ran on an a100 80gb gpu.
Configs:
```
    # Run a bunch of experiments
    batch_sizes = [8, 16, 32]
    num_heads = [16, 32]
    max_seq_lens = [15, 64, 128, 512, 555, 1024]
    embed_dims = [32, 64, 128]
    dtypes = [torch.float16, torch.bfloat16, torch.float32]
    pad_percentages = [None]
    backends = [SDPBackend.EFFICIENT_ATTENTION, SDPBackend.MATH]
    run_backward = True
    attn_mask = True
```
    
   The function calls `sdpa(input**).sum().backward()`.
   
   I calculated the geomean speedup of the efficient attention path of the math path for all these configs:
   `Geomean Speedup: 1.977`
   
An example comparision with batchsize = 8, num_heads = 32, embed_dim = 64, and dtype = torch.float16:
attn_mask_compare_bsz_8_num_heads_32_embed_dim_64_dtype_fp16

 This was done using the current state of the branch where we force alignment of mask when the last dim is not divisible by 16, which shows up in seq_len = 15 and 555 case.

The full data can be found here:

   
[attn_mask_sweep.csv](https://github.com/pytorch/pytorch/files/11962399/attn_mask_sweep.csv)

File: aten/src/ATen/native/transformers/attention.cpp [hunk 4]
```diff
@@ -558,9 +561,67 @@ inline void validate_sdpa_input(
...
+at::Tensor pad_bias(const at::Tensor& attn_bias) {
+  int align_to = 16;
+  if (attn_bias.size(-1) % align_to == 0) {
+    return attn_bias;
+  }
+  int pad_count = align_to - (attn_bias.size(-1) % align_to);
+  auto padded_bias = at::pad(attn_bias, {0, pad_count});
+  return padded_bias.slice(-1, 0, attn_bias.size(-1));
+}
+// Hmmm bad optional access syms like the above needs to be symintified
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/attention.cpp:607]: lol seems*

**Comment by drisspg** [comment]
```
[2023-07-07 14:13:10,038] torch.fx.experimental.symbolic_shapes: [WARNING] Ignored guard Eq(256, s3) == False, this could result in accuracy problems
[2023-07-07 14:13:10,042] torch.fx.experimental.symbolic_shapes: [WARNING] Ignored guard Ne(256, s3) == True, this could result in accuracy problems
Using FallbackKernel: aten._scaled_dot_product_efficient_attention_backward
[2023-07-07 14:13:10,282] torch.fx.experimental.symbolic_shapes: [WARNING] Ignored guard s0*s1*s3 < 2147483648 == True, this could result in accuracy problems
[2023-07-07 14:13:10,291] torch.fx.experimental.symbolic_shapes: [WARNING] Ignored guard 65536*s0 + 256*s1 + s3 - 65792 < 2147483648 == True, this could result in accuracy problems
```

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -14200,7 +14200,7 @@
     CUDA: _efficient_attention_forward
   tags: nondeterministic_seeded
 
-- func: _efficient_attention_backward(Tensor grad_out_, Tensor query, Tensor key, Tensor value, Tensor? bias, Tensor out, Tensor? cu_seqlens_q, Tensor? cu_seqlens_k, int max_seqlen_k, int max_seqlen_q, Tensor logsumexp, float dropout_p, Tensor philox_seed, Tensor philox_offset, int custom_mask_type, *, float? scale=None, int? num_splits_key=None) -> (Tensor, Tensor, Tensor, Tensor)
+- func: _efficient_attention_backward(Tensor grad_out_, Tensor query, Tensor key, Tensor value, Tensor? bias, Tensor out, Tensor? cu_seqlens_q, Tensor? cu_seqlens_k, int max_seqlen_k, int max_seqlen_q, Tensor logsumexp, float dropout_p, Tensor philox_seed, Tensor philox_offset, int custom_mask_type, bool bias_requires_grad, *, float? scale=None, int? num_splits_key=None) -> (Tensor, Tensor, Tensor, Tensor)
```
Comment (drisspg) [inline aten/src/ATen/native/native_functions.yaml:14203]: extra arg makes it easier to right meta func

File: aten/src/ATen/native/transformers/attention.cpp [hunk 6]
```diff
@@ -538,9 +541,53 @@ inline void validate_sdpa_input(
...
+c10::optional<Tensor> convert_boolean_attn_mask(const c10::optional<Tensor>& attn_mask, caffe2::TypeMeta dtype) {
+  // Pass through
+  if(!attn_mask.has_value()){
+    return c10::nullopt;
+  }
+  // Convert boolean mask to additive mask; need to invert mask to indicate what
+  // to mask *out*.
+  if (attn_mask->dtype() == at::kBool) {
+    auto new_attn_mask = at::zeros_like(attn_mask.value(), dtype);
+    // TODO Use the max type of the input and output
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/attention.cpp:565]: Maybe todo

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 7]
```diff
@@ -524,7 +524,11 @@ std::tuple<Tensor, Tensor> native_multi_head_attention_cuda(
     // strides from packed projection for nested tensors when seq_len is 1 will be
     // and will trigger a contiguous call in the kernel, so we prevent this
     bool no_seq_len_1_nested = query.is_nested() ? check_for_seq_len_1_nested_tensor(kernel_params, false) : true;
-    if (no_seq_len_1_nested &&
+    // The API for transfomer_encoder is a mask of shape (Batch_Size, Seq_len_q)
+    // For mem-eff attention this will cause the expand call to
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention.cu:528]: fix comment, `to error`

File: aten/src/ATen/native/transformers/cuda/attention_backward.cu [hunk 8]
```diff
@@ -187,14 +188,15 @@ _efficient_attention_backward(
...
-  const bool bias_requires_grad = bias.has_value() && bias->requires_grad();
-
   at::Tensor grad_q, grad_k, grad_v, grad_bias;
   grad_q = at::empty(query.sizes(), query.options());
   grad_k = at::empty(key.sizes(), key.options());
   grad_v = at::empty(value.sizes(), value.options());
 
   if (bias_requires_grad) {
+    // Alignment has already been enforced
+    // grad_bias = at::empty_strided((*bias).sizes(), (*bias).strides(), (*bias).options());
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention_backward.cu:198]: can remove this line of code

File: test/test_transformers.py [hunk 9]
```diff
@@ -1237,22 +1237,18 @@ def test_invalid_fused_inputs_invalid_dtype(self, device, kernel: SDPBackend):
...
     def test_invalid_fused_inputs_attn_mask_present(self, device, kernel: SDPBackend):
         with sdp_kernel(**backend_map[kernel]):
             # Failures for unsupported SDP args
             size = (2, 2, 3, 16)
             make_tensor = partial(rand_sdpa_tensor, type="dense", device=device, dtype=torch.float16)
             q, k, v = make_tensor(size), make_tensor(size), make_tensor(size)
             # Non-None attention mask
+            mask = torch.ones((2, 2, 3, 3), device=device, dtype=q.dtype)
             self.assertRaises(RuntimeError, lambda: torch.nn.functional.scaled_dot_product_attention(
-                q, k, v, torch.ones_like(q), 0.0, False))
```
Comment (drisspg) [inline test/test_transformers.py:1255]: this mask was just wrong

File: test/test_transformers.py [hunk 10]
```diff
@@ -1953,7 +1949,122 @@ def _get_mem_eff_drop_mask(batch_size, n_heads, q_len, kv_len, p, seed, offset,
...
+    @parametrize("batch_size", [1, 8])
+    @parametrize("seq_len_q", [4, 8, 64, 128, 256, 312, 512, 1024, 2048] if SM80OrLater else [4, 8, 64, 128, 152, 256, 512])
+    @parametrize("seq_len_k", [4, 8, 64, 65, 128, 256, 408, 512, 1024, 2048] if SM80OrLater else [4, 8, 37, 64, 128, 256, 512])
+    @parametrize("head_dim", [8, 16, 32, 64, 72, 96, 128] if SM80OrLater else [8, 16, 32, 64])
+    @parametrize("is_causal", [False])
+    @parametrize("dropout_p", [0.0, 0.22])
+    @parametrize("dtype", [torch.float16, torch.bfloat16, torch.float32] if
+                 SM80OrLater else [torch.float16, torch.float32])
+    @parametrize("scale", [None, "l1"])
+    def test_mem_efficient_attention_attn_mask_vs_math_ref_grads(self, device, batch_size: int, seq_len_q: int,
```
Comment (drisspg) [inline test/test_transformers.py:1963]: can be merged into existing test, but makes it even more monolithic

**Review by drisspg — COMMENTED** [review]
Not bad

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py [hunk 11]
```diff
@@ -267,14 +267,14 @@
     ("aten::_scaled_dot_product_attention", datetime.date(2023, 8, 1)),
     ("aten::_chunk_grad_outputs_efficient_attention", datetime.date(2023, 8, 1)),
     ("aten::_scaled_dot_product_flash_attention", datetime.date(2023, 5, 15)),
-    ("aten::_scaled_dot_product_efficient_attention", datetime.date(2023, 7, 1)),
-    ("aten::_scaled_dot_product_efficient_attention_backward", datetime.date(2023, 7, 1)),
+    ("aten::_scaled_dot_product_efficient_attention", datetime.date(2023, 7, 15)),
```
Comment (cpuhrsch) [inline test/forward_backward_compatibility/check_forward_backward_compatibility.py:270]: Do these times need to be updated at this point?

File: aten/src/ATen/native/transformers/attention.cpp [hunk 12]
```diff
@@ -591,16 +639,25 @@ Tensor scaled_dot_product_attention(
...
           (query_.requires_grad() || key.requires_grad() ||
            value.requires_grad());
+      if (attn_mask.has_value()) {
+        // Expand to 4d case
+        attn_mask = attn_mask.value().expand_symint(
+            {query_.sym_size(0),
+             query_.sym_size(1),
+             query_.sym_size(2),
+             key.sym_size(2)});
+        attn_mask = pad_bias(attn_mask.value());
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/attention.cpp:649]: So when the attention mask is a bool, you'll have allocated a new float attention mask in `convert_boolean_attn_mask` and then subsequently will reallocate to pad it out to a larger one if it's not a multiple of 16. 

Seems like an opportunity to allocate the right size right away if you need padding, i.e. extend `convert_boolean_attn_mask` to also do the padding. You'll know at `convert_boolean_attn_mask` time what backend you're about to choose, so it doesn't require more special casing.

File: aten/src/ATen/native/transformers/attention.cpp [hunk 13]
```diff
@@ -639,22 +696,15 @@ std::tuple<Tensor, Tensor> _scaled_dot_product_attention_math(
         // Replace attn_mask with causal mask; lower triangular elements take part in attention.
         const auto L = query.sym_size(-2), S = key.sym_size(-2);
         attn_mask = at::ones_symint({L, S}, query.options().dtype(at::kBool)).tril();
+        attn_mask = convert_boolean_attn_mask(attn_mask, query.dtype());
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/attention.cpp:699]: nit: Similar story here where the mask was just allocated and now getting reallocated again. [`triu`](https://pytorch.org/docs/master/generated/torch.triu.html#torch.triu) should also work for float so if `convert_boolean_attn_mask` returns a fully `-inf`'d attn_mask you could use triu to cut off the top part.

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 14]
```diff
@@ -1045,8 +1049,14 @@ std::tuple<at::Tensor, at::Tensor, Tensor, Tensor> _efficient_attention_forward(
 
       // assign strides for bias, viewed as
       // (batch_sz, n_heads, n_queries, n_keys)
-      const at::Tensor bias_4d_view =
-          get_bias_4d_view(*bias, B, num_heads, M, N);
+      // We make sure to expand prior to calling the kernel
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/cuda/attention.cu:1052]: Are there any checks in place at kernel level that'll throw an exception if the bias' shape isn't a multiple of 16?

File: aten/src/ATen/native/transformers/attention.cpp [hunk 15]
```diff
@@ -591,16 +639,25 @@ Tensor scaled_dot_product_attention(
       bool compute_logsumexp =
           (query_.requires_grad() || key.requires_grad() ||
            value.requires_grad());
+      if (attn_mask.has_value()) {
+        // Expand to 4d case
+        attn_mask = attn_mask.value().expand_symint(
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/attention.cpp:644]: Random thought, but I saw that the kernel takes into account the strides of the bias. Is it really always better to just expand for the 2D case or is it maybe better to also add a `.contiguous` after? I could see a small Tensor being this much in demand being harmful to performance.

At the same time, it saves a ton of memory to not replicate the mask for each batch entry. I'm mostly curious, but I could see that maybe in some cases it'd be better to replicate the mask per head at least? Might be worth some follow up investigation / performance tuning.

File: aten/src/ATen/native/transformers/cuda/attention.cu (hunk 14 above)
Comment (drisspg) [reply aten/src/ATen/native/transformers/cuda/attention.cu:1052]: Yeah, this torch_check will pop: https://github.com/pytorch/pytorch/blob/14304afd76a6e8c383fe833d1f28122d45ae99eb/aten/src/ATen/native/transformers/cuda/mem_eff_attention/kernel_forward.h#L583

**Review by cpuhrsch — APPROVED** [review]

**Comment by fxmarty** [comment]
@drisspg looks very good. I guess enabling the 1MHK mode would be even faster for the padding case? Do you know if it is planned? `_scaled_dot_product_efficient_attention_cuda` always passes `c10::nullopt` for `seqstart_q` & `seqstart_k`, thus I believe the full sequence is always used.

**Comment by drisspg** [comment]
> @drisspg looks very good. I guess enabling the 1MHK mode would be even faster for the padding case? Do you know if it is planned? `_scaled_dot_product_efficient_attention_cuda` always passes `c10::nullopt` for `seqstart_q` & `seqstart_k`, thus I believe the full sequence is always used.
> 
> Maybe 1MHK can be enabled with nestedtensor? Though this commit [facebookresearch/xformers@cfea89f#diff-6ffc547e102d4e367e017498c0697c9be0b1901c9f940fef64a2a5c49b613c0b](https://github.com/facebookresearch/xformers/commit/cfea89fab5e42a99f961d445ac2b06d4b43126ce#diff-6ffc547e102d4e367e017498c0697c9be0b1901c9f940fef64a2a5c49b613c0b) seem to hint that flash v2 is always faster than the other xformers paths.

Yup, thats exactly what are going to look into

**Pull request closed without merging**